main.tf

```diff
@@ -177,12 +176,63 @@ resource "github_branch_default" "default" {
+
+resource "github_branch_protection" "this" {
+  for_each = var.branch_protections_v4
+
+  # ensure we have all members and collaborators added before applying
+  # any configuration for them
+  depends_on = [
+    github_repository_collaborator.collaborator,
+    github_team_repository.team_repository,
+    github_team_repository.team_repository_by_slug,
+    github_branch.branch,
+  ]
+
+  repository_id                   = github_repository.repository.node_id
+  pattern                         = each.key
+  allows_deletions                = coalesce(each.value.allows_deletions, false)
+  allows_force_pushes             = coalesce(each.value.allows_force_pushes, false)
+  blocks_creations                = coalesce(each.value.blocks_creations, false)
+  enforce_admins                  = coalesce(each.value.enforce_admins, false)
+  push_restrictions               = coalesce(each.value.push_restrictions, [])
+  require_conversation_resolution = coalesce(each.value.require_conversation_resolution, false)
+  require_signed_commits          = coalesce(each.value.require_signed_commits, false)
+  required_linear_history         = coalesce(each.value.required_linear_history, false)
+
+  dynamic "required_pull_request_reviews" {
+    for_each = each.value.required_pull_request_reviews != null ? [true] : []
+
+    content {
+      dismiss_stale_reviews           = coalesce(each.value.required_pull_request_reviews.dismiss_stale_reviews, false)
+      dismissal_restrictions          = coalesce(each.value.required_pull_request_reviews.dismissal_restrictions, [])
+      pull_request_bypassers          = coalesce(each.value.required_pull_request_reviews.pull_request_bypassers, [])
+      require_code_owner_reviews      = coalesce(each.value.required_pull_request_reviews.require_code_owner_reviews, false)
+      required_approving_review_count = coalesce(each.value.required_pull_request_reviews.required_approving_review_count, 0)
+      restrict_dismissals             = length(coalesce(each.value.required_pull_request_reviews.dismissal_restrictions, [])) > 0
+    }
+  }
+
+  dynamic "required_status_checks" {
+    for_each = each.value.required_status_checks != null ? [true] : []
+
+    content {
+      strict   = coalesce(each.value.required_status_checks.strict, false)
+      contexts = coalesce(each.value.required_status_checks.contexts, [])
+    }
+  }
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# v3 Branch Protection
 # https://registry.terraform.io/providers/integrations/github/latest/docs/resources/branch_protection_v3
 # ---------------------------------------------------------------------------------------------------------------------
 
 resource "github_branch_protection_v3" "branch_protection" {
-  count = length(local.branch_protections)
+  count = var.branch_protections_v4 == null ? length(local.branch_protections_v3) : 0
```

Comment:
I've been able to support v3 & v4 side by side using the following check here : `count = length(coalesce(var.branch_protections_v4, {})) == 0 ? length(local.branch_protections_v3) : 0
`